libstuff/STCPManager.cpp

```diff
@@ -8,218 +8,196 @@
+            // Always send/recv (see Socket::CONNECTED, above)
+            // **FIXME: Add timeout.
+            bool aliveAfterRecv = socket.recv();
+            bool aliveAfterSend = socket.send();
+            if (!aliveAfterSend || (!aliveAfterRecv && socket.sendBufferEmpty())) {
+
+                // Did we send everything?  (Technically this the send buffer could be empty and we still haven't
+                // sent everything -- SSL buffers internally, so we should check that buffer.  But odds are it sent fine.)
+                if (socket.sendBufferEmpty()) {
+                    SDEBUG("Graceful shutdown of SSL socket '" << socket.addr << "'");
+                } else {
+                    SWARN("Dirty shutdown of SSL socket '" << socket.addr << "' (" << socket.sendBufferCopy().size()
+                                                           << " bytes remain)");
                 }
-            } else {
-                // Not SSL -- only send if we have something to send
-                if (!socket->sendBufferEmpty()) {
-                    // Still have something to send -- try to send it.
-                    if (!socket->send()) {
-                        // Done trying to send
-                        SHMMM("Unable to finish sending to '" << socket->addr << "' on shutdown, clearing.");
-                        ::shutdown(socket->s, SHUT_RDWR);
-                        socket->setSendBuffer("");
-                    }
+                socket.shutdown(Socket::CLOSED);
+            }
+        } else {
+            // Not SSL -- only send if we have something to send
+            if (!socket.sendBufferEmpty()) {
+                // Still have something to send -- try to send it.
+                if (!socket.send()) {
+                    // Done trying to send
+                    SHMMM("Unable to finish sending to '" << socket.addr << "' on shutdown, clearing.");
+                    socket.shutdown();
+                    socket.setSendBuffer("");
                 }
+            }
 
-                // Are we done sending?
-                // **FIXME: Add timeout
-                if (socket->sendBufferEmpty()) {
-                    // Wait for the other side to shut down
-                    if (!socket->recv()) {
-                        // Done shutting down
-                        SDEBUG("Graceful shutdown of socket '" << socket->addr << "'");
-                        socket->state.store(Socket::CLOSED);
-                        ::shutdown(socket->s, SHUT_RDWR);
-                    }
+            // Are we done sending?
+            // **FIXME: Add timeout
+            if (socket.sendBufferEmpty()) {
+                // Wait for the other side to shut down
+                if (!socket.recv()) {
+                    // Done shutting down
+                    SDEBUG("Graceful shutdown of socket '" << socket.addr << "'");
+                    socket.shutdown(Socket::CLOSED);
                 }
             }
```

Comment:
Is it possible to reach this block after the above block runs? i.e. Send buffer was not empty, so we sent, and then the send buffer is still not empty. If so, what happens? Or does this whole method get called in a loop that will catch it next time around?

Reply:
Yes, it's possible, we will run the whole method again, it does loop.